lms/resources/_js_config/__init__.py

```diff
@@ -510,28 +511,30 @@ def _canvas_sync_api(self):
 
     def _blackboard_sync_api(self):
         req = self._request
-        return {
-            "authUrl": req.route_url("blackboard_api.oauth.authorize"),
-            "path": req.route_path("blackboard_api.sync"),
-            "data": {
-                "lms": {
-                    "tool_consumer_instance_guid": req.params[
-                        "tool_consumer_instance_guid"
-                    ],
-                },
-                "course": {
-                    "context_id": req.params["context_id"],
-                },
-                "assignment": {
-                    "resource_link_id": req.params["resource_link_id"],
-                },
-                "group_info": {
-                    key: value
-                    for key, value in req.params.items()
-                    if key in GroupInfo.columns()
+        return APISyncBlackboardSchema(req).load(
```

Comment:
I like the idea. I think Marshmallow schemas have two methods `dump()` and `load()`:

1. `dump()` serializes a model object (possibly just a dict) into a dict. See <https://marshmallow.readthedocs.io/en/stable/quickstart.html#serializing-objects-dumping>.

2. `load()` deserializes a dict into a model object (possibly just a dict): https://marshmallow.readthedocs.io/en/stable/quickstart.html#deserializing-objects-loading

So I think if we want to use `APISyncBlackboardSchema` to serialize the data dict to send to the frontend we want to use `APISyncBlackboardSchema.load()` for that.

I think the schema is only about the request body (the `data` part) so I think it'd be:

```python
return {
    "authUrl": ...,
    "path": ...,
    "data": APISyncBlackboardSchema(req).dump(...),
}
```

Reply:
I'm not sure that dumping actually requires the `request` object, so I think you should be able to do `APISyncBlackboardSchema().dump({...})`. But this might be an existing issue with `PyramidRequestSchema`, the `request` should be an argument to `parse()` rather than to `__init__()`

Reply:
Actually, rather than having `APISyncBlackboardSchema` just validate the data to be serialized, maybe it should generate it? Something like:

```python
>>> APISyncBlackboardSchema.dump(request)
{
    "lms": {
        "tool_consumer_instance_guid": "...",
        ...,
    },
    ...
}
```

This would mean that if we want to change the sync API params (e.g. to add a new param) we only need to change it in one place: in the `APISyncBlackboardSchema` class, since the schema class is responsible for both generating the params that we send to the frontend and validating the params that the frontend sends back to us.

Some of the authentication-related schema classes do something similar. For example `BearerTokenSchema` both serializes a bearer token from an `LTIUser` and deserializes a bearer token into an `LTIUser`. `OAuthCallbackSchema` also does something similar